src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -266,38 +266,101 @@ int ReplicationStateMachine::read_snapshot_data(std::shared_ptr< homestore::snap
 
 void ReplicationStateMachine::write_snapshot_data(std::shared_ptr< homestore::snapshot_context > context,
                                                   std::shared_ptr< homestore::snapshot_data > snp_data) {
-    // RELEASE_ASSERT(context != nullptr, "Context null");
-    // RELEASE_ASSERT(snp_data != nullptr, "Snapshot data null");
-    // auto s = dynamic_pointer_cast< homestore::nuraft_snapshot_context >(context)->nuraft_snapshot();
-    // int64_t obj_id = snp_data->offset;
-    // uint64_t shard_seq_num = obj_id >> 16;
-    // uint64_t batch_number = obj_id & 0xFFFF;
-    //
-    // auto log_str = fmt::format("group={} term={} lsn={} shard_seq={} batch_num={} size={}",
-    //                            boost::uuids::to_string(repl_dev()->group_id()), s->get_last_log_term(),
-    //                            s->get_last_log_idx(), shard_seq_num, batch_number, snp_data->blob.size());
-    //
-    // if (snp_data->is_last_obj) {
-    //     LOGD("Write snapshot reached is_last_obj true {}", log_str);
-    //     return;
-    // }
-    //
-    // if (obj_id == 0) {
-    //     snp_data->offset = 1 << 16;
-    //     // TODO add metadata.
-    //     return;
-    // }
-    //
-    // auto snp = GetSizePrefixedResyncBlobDataBatch(snp_data->blob.bytes());
-    // // TODO Add blob puts
-
-    // if (snp->end_of_batch()) {
-    //     snp_data->offset = (shard_seq_num + 1) << 16;
-    // } else {
-    //     snp_data->offset = (shard_seq_num << 16) | (batch_number + 1);
-    // }
-    //
-    // LOGT("Read snapshot num_blobs={} end_of_batch={} {}", snp->data_array()->size(), snp->end_of_batch(), log_str);
+    RELEASE_ASSERT(context != nullptr, "Context null");
+    RELEASE_ASSERT(snp_data != nullptr, "Snapshot data null");
+    auto s = dynamic_pointer_cast< homestore::nuraft_snapshot_context >(context)->nuraft_snapshot();
+    auto obj_id = objId(static_cast< snp_obj_id_t >(snp_data->offset));
+    auto r_dev = repl_dev();
+
+    auto log_suffix = fmt::format("group={} term={} lsn={} shard={} batch_num={} size={}",
+                                  uuids::to_string(r_dev->group_id()), s->get_last_log_term(), s->get_last_log_idx(),
+                                  obj_id.shard_seq_num, obj_id.batch_id, snp_data->blob.size());
+
+    if (snp_data->is_last_obj) {
+        LOGD("Write snapshot reached is_last_obj true {}", log_suffix);
+        return;
+    }
+
+    // Check if the snapshot context is same as the current snapshot context.
+    // If not, drop the previous context and re-init a new one
+    if (m_snp_rcv_handler && m_snp_rcv_handler->snp_lsn_ != context->get_lsn()) { m_snp_rcv_handler.reset(nullptr); }
```

Comment:
I don't think I follower in which case we should reuse this m_snp_rcv_handler. Every snapshot context will have distinguished lsn, no? Or there is a case two snapshots will have the same lsn?